### ISSUE TITLE
Why is there no `mtrand.pxd`?

### ISSUE THREAD

**Issue by jbn** [numpy/numpy#7861]
A thread from 2010 in [`python-cython-user` articulated my exact problem](http://osdir.com/ml/python-cython-user/2010-10/msg00119.html). Namely, I'm doing (things like) MCMC, and I want to use numpy's MT implementation. Mostly, I do this because I don't want multiple PRNG objects floating around my implementation. And, I'm using the `RandomState` distributions in places other than the Cythonized code that needs access to the PRNG. 

I searched the historical open and closed issues, but I never found a solution. Is there any reason, obvious or otherwise, why numpy does not include a `.pxd` counter-part to `mtrand.pyx`? If not, I'll do this work and submit a PR. But, considering I often encounter this need, I'm assuming a well-justified reason for its absence.

Thanks!

**Comment by njsmith** [comment]
Probably the first reason that there is no mtrand.pxd is that most work on mtrand predated the invention of pxd files.

That said, it's not clear to me what value you're hoping to get out of having an mtrand.pxd. Currently, all the mtrand operations are defined in terms of numpy array objects and dtype objects and so forth, so there isn't really any better way to call them than through the existing python interface.

Can you give some more details? Where exactly are you hoping to call mtrand from (cython? C?), what sort of interface do you want, and what problems does this interface solve compared to the current one?

**Comment by jbn** [comment]
Pseudo-code:

```cython
from numpy.random cimport RandomState  # Uh-oh.
from numpy.random cimport rk_state  # Doable: `cdef extern from "randomkit.h"`

cdef class BanditLikeAgent:
    cpdef readonly object internal_state

    # Here is the basic need. I want to offer a client-facing API that just 
    # asks for `numpy.random` or an instantiated `RandomState`. But, 
    # internally, I'm mostly sampling from `rk_state` and the associated 
    # distribution functions. 
    cpdef void step(self, RandomState random_state):
        cdef rk_state* rng = random_state.internal_state
        
        while True:
            x = rk_double(rng)
            
            # Do stuff...
            sub_routine(rng)
            
            # Do more stuff...
```

**Comment by njsmith** [comment]
The big request here then is to define a steeper, supported, C level api for random number generation (including multiple distributions). This is a pretty big request -- currently all those rk_ functions are private implementation details, that can change at whim (and probably will in the near-ish future as we figure out how to support multiple generator backends). We should get @rkern's thoughts too for sure, but my first reaction is that this would first require largely rewriting the internal C api into something that was more abstracted from the implementation.

**Comment by jbn** [comment]
Yes, it is non-trivial. 

I guess I am too biased to judge whether it is a feature many people want. For me, **having C-access to `RandomState`, `rk_state`, and `rk_double` solves 95% of my use cases**. I suspect the same is true for most other people doing ABM work. On the other hand, even people doing MCMC probably don't need this. They have PyMC or PyStan or something, and rarely need to directly draw from the PRNG.

**Comment by bashtage** [comment]
@jbn Does numba (http://numba.pydata.org/) work for your problem?  They have shimmed numpy.random to work in their `nopython` mode so that it can be used in tight loops withouth Python object overhead.

**Comment by rkern** [comment]
Incorrectly, yes.

**Comment by bashtage** [comment]
No warranties or endorsements, either expressed or implied.

**Comment by njsmith** [comment]
I guess there isn't any reason in principle why we couldn't have a C API function that took a RandomState* and returned uniformly sampled doubles. I think we'd want to make it part of the existing C API export table instead of relying on cython's api export machinery, just to keep down the surface area of stuff we're committing to supporting, and this would be technically somewhat tricky since it would require some way for multiarray.so to get at the internals of mtrand.pyx. But at least the public api seems reasonable if someone can figure out how to make the internals work.

**Comment by jbn** [comment]
I've been trying to avoid `numba` for my current project. I can't articulate why though. It's not privileged information, I simply don't have a reasoned argument. I'll try to investigate how well `nopython` would work in my case next weekend. (Unless, @rkern, are you saying it's dangerous?)

**Comment by bashtage** [comment]
It isn't completely unreasonable to limit numba exposure given the number of regressions across the monthly releases that are coming, although it has really expanded its capabilities this year including common matrix ops.

**Comment by rkern** [comment]
@jbn numba is not dangerous per se, just that the way it shims into `numpy.random` is unsupported and abuses implementation details. They have been invited to help work on exposing a stable C API for `numpy.random`, as you are asking for, but no one has stepped up to do that work yet.

**Comment by mattip** [comment]
If cffi ever becomes an acceptable dependency, we could rewrite the random module as as shared object  [which was done for PyPY](https://bitbucket.org/pypy/numpy/src/master/numpy/random/_mtrand_build.py), and wrap that with a cffi module

**Comment by bashtage** [comment]
I've implemented some experiments along this line (both CFFI and Ctypes) in randomgen:

See, e.g., 

https://github.com/bashtage/randomgen/blob/master/randomgen/examples/numba/extending_distributions.py

and 

https://bashtage.github.io/randomgen/extending.html

**Comment by jbn** [comment]
Yes, in a lot of cases, I think so! I'm going to create a branch of my dissertation code and see if theirs a nice numba performance boost next week.

**Comment by mattip** [comment]
@jbn, any progress here?

**Comment by jbn** [comment]
@mattip None from me.

**Comment by mattip** [comment]
So should we close this?

**Comment by bashtage** [comment]
I think it should be closed when #13163 gets merged

**Comment by mattip** [comment]
The new API's cython support is [documented](http://www.numpy.org/devdocs/reference/random/extending.html#cython). Closing.
